src/cargo/core/compiler/timings.rs

```diff
@@ -90,11 +91,75 @@ struct UnitTime {
     unlocked_units: Vec<Unit>,
     /// Same as `unlocked_units`, but unlocked by rmeta.
     unlocked_rmeta_units: Vec<Unit>,
+    /// Individual compilation section durations, gathered from `--json=timings`.
     /// IndexMap is used to keep original insertion order, we want to be able to tell which
     /// sections were started in which order.
     sections: IndexMap<String, CompilationSection>,
 }
 
+const FRONTEND_SECTION_NAME: &str = "Frontend";
+const CODEGEN_SECTION_NAME: &str = "Codegen";
+
+impl UnitTime {
+    fn aggregate_sections(&self) -> AggregatedSections {
+        let end = self.duration;
+
+        if !self.sections.is_empty() {
+            // We have some detailed compilation section timings, so we postprocess them
+            // Since it is possible that we do not have an end timestamp for a given compilation
+            // section, we need to iterate them and if an end is missing, we assign the end of
+            // the section to the start of the following section.
```

Comment:
Are sections non-overlapping, allowing this to work?

Reply:
Currently, yes. If it changes in the future, we'd need to of course adapt to that.